File: src/207.课程表.rs
/*
 * @lc app=leetcode.cn id=207 lang=rust
 *
 * [207] 课程表
 */
struct Solution;
// @lc code=start
use std::collections::VecDeque;
use std::iter::FromIterator;
impl Solution {
    pub fn can_finish(n: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        let n = n as usize;
        // form graph
        let mut next = vec![Vec::new(); n];
        let mut in_degree = vec![0; n];
        for edge in prerequisites.iter() {
            // (target <- course)
            next[edge[1] as usize].push(edge[0] as usize);
            in_degree[edge[0] as usize] += 1;
        }
        let mut q = VecDeque::from_iter((0..n).filter(|&i| in_degree[i] == 0));
        let mut finished = 0;
        while q.len() > 0 {
            let course = q.pop_front().unwrap();
            // mark course as finished
            finished += 1;
            // mark prerequisites as finished
            for &target in next[course].iter() {
                in_degree[target] -= 1;
                if in_degree[target] == 0 {
                    q.push_back(target);
                }
            }
        }

        finished == n
    }
}
// @lc code=end
mod utils;
#[test]
fn test_solution() {
    macro_rules! test {
        ($n:expr, $p:tt, $ans:expr) => {
            assert_eq!(
                Solution::can_finish($n, vec2d!$p),
                $ans
            );
        }
    };
    test!(2, [[1, 0]], true);
    test!(2, [[1, 0], [0, 1]], false);
}

File: src/207.课程表.py
#
# @lc app=leetcode.cn id=207 lang=python3
#
# [207] 课程表
#
from typing import List
# @lc code=start
from queue import Queue


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        edges = [[] for _ in range(numCourses)]
        in_degree = [0 for _ in range(numCourses)]
        for b, a in prerequisites:
            edges[a].append(b)
            in_degree[b] += 1
        #
        zero_in_degree = Queue()
        for i, degree in enumerate(in_degree):
            if degree == 0:
                zero_in_degree.put(i)
        #
        finished_courses = 0
        while zero_in_degree.qsize() > 0:
            finish = zero_in_degree.get()
            # finish course
            finished_courses += 1
            for next_course in edges[finish]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    zero_in_degree.put(next_course)
        return finished_courses == numCourses


# @lc code=end
if __name__ == "__main__":
    f = Solution().canFinish
    assert f(2, [[1, 0]])
    assert not f(2, [[1, 0], [0, 1]])
